### PR TITLE
Add TransformerType protocol

### PR DESCRIPTION
**Pull request by chuganzy**
Now, we can define operator function like below.

``` swift
public func <| <T: TransformerType where T.From: Decodable, T.From.DecodedType == T.From>(e: Extractor, pair: (KeyPath, T) ) throws -> T.To {
    return try pair.1.apply(e <| pair.0)
}
```

### CODE REVIEW

File: Sources/Transformer.swift [hunk 1]
```diff
@@ -6,44 +6,20 @@
...
+    public typealias From = F
+    public typealias To = T
+    private let transform: From -> To?
 
-    public init(_ transform: From throws -> To) {
+    public init(_ transform: From -> To?) {
         self.transform = transform
     }
 
     /// - Throws: DecodeError
```
Comment (chuganzy) [inline Sources/Transformer.swift:18]: It depends on the `transform` block, so change `transform` type from `From throws -> To` to `From -> To?`.

File: Tests/Himotoki/TransformerTest.swift [hunk 2]
```diff
@@ -30,7 +26,7 @@ private struct URLsByTransformer: Decodable {
         let URLTransformer = Transformer(toURL)
 
         return self.init(
-            value: try Transformer { try toURL($0) }.apply(e <| "value"),
```
Comment (ikesyo) [inline Tests/Himotoki/TransformerTest.swift:33]: This is verifying `Transformer` can be initialized with trailing closure.

File: Sources/Transformer.swift [hunk 3]
```diff
@@ -6,44 +6,20 @@
...
 
-public struct Transformer<From, To> {
-    private let transform: From throws -> To
+public struct Transformer<F, T>: TransformerType {
+    public typealias From = F
+    public typealias To = T
+    private let transform: From -> To?
 
-    public init(_ transform: From throws -> To) {
+    public init(_ transform: From -> To?) {
```
Comment (ikesyo) [inline Sources/Transformer.swift:14]: This is a breaking API change, so we can't accept the change into `master` currently.

**Comment by ikesyo** [comment]
Thanks for the pull request! :pray: 

However, this is a breaking API change and this loses the flexibility that the users can inform the reason why the transformation failed. I think this is not a right direction we should go in.

Thank you anyway @hoppenichu!

File: Tests/Himotoki/TransformerTest.swift (hunk 2 above)
Comment (chuganzy) [inline Tests/Himotoki/TransformerTest.swift:33]: I see!

**Comment by hoppenichu** [comment]
OK, as I expected this will be closed.

I have few questions implementing this.

1. What is the purpose of adding `Transformer` to Himotoki? It is completely standalone and separatable: users can define same kind of function.

2. `/// - Throws: DecodeError` in `Transform` are wrong. Then what is your plan to fix them?

**Pull request closed without merging**